posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd

```diff
@@ -0,0 +1,372 @@
+  theme(text=element_text(size=15))+
+  ggtitle(sprintf("Reshaping from wide to long panda & data.table \nover real numbers, N times", n.folds))+
+  scale_x_log10("N = number of Mean,SD,Length to compute")+
+  scale_y_log10("Computation time (seconds)
+median line, min/max band
+over 10 timings")+
+  facet_null()+
+  scale_fill_manual(values=ml.colors)+
+  scale_color_manual(values=ml.colors)
+```
+
+```{r, warning=FALSE, message=FALSE}
+plot(ml.wide2long.pd)
+```
+
+When converting data from wide to long format, as shown in the above plot, data.table's melt() function efficiently gathers multiple columns into key-value pairs, allowing for easy transformation of wide data into a longer, more structured format. 
+
+
+
+## B. long to wide reshape
+
+In data.table, the data.table::dcast() function is often used to convert data from a long format to a wide format, and in pandas, the pandas::pivot_table() function is used to convert data from a long format to a wide format.
+
+## Comparison code
+
+## data.table::dcast() is faster
+
+```{r,warning=FALSE,message=FALSE,results='hide'}
+
+ml.colors <- c(
+  "data.table::dcast" = "#D6604D",
+  "pandas::pivot_table" = "#BF812D"
+)
+
+n.folds <- 10
+n.rows <- 100
+seconds.limit <- 1
+
+ml.long2wide.atime <- atime::atime(
+  N=as.integer(10^seq(2, 7, by=0.2)),
+  setup={
+    df <- data.frame(
+      id = rep(1:N, each = 2),
+      category = rep(c("A", "B"), N),
+      value = rnorm(2 * N)
+      )
+    },
+  seconds.limit= seconds.limit,
+  "data.table::dcast" = {
+    data.table::dcast(data.table(df), id ~ category, value.var = "value")
+  },
+  "pandas::pivot_table" = {
+    py_df <- reticulate::r_to_py(df)
+    pd <- import("pandas")
+    pd$pivot_table(py_df, values = "value", index = "id", columns = "category")
+  }
+  )
+
+```
+
+
+
+```{r,warning=FALSE,message=FALSE}
+ml.long2wide.refs <- atime::references_best(ml.long2wide.atime)
+plot(ml.long2wide.refs)
+```
+same as all the above, data::dcast performs better in terms of complexity notations 
+
+```{r,warning=FALSE,message=FALSE}
+ml.long2wide.pred <- predict(ml.long2wide.refs)
+plot(ml.long2wide.pred)
+```
+
+
+```{r,warning=FALSE,message=FALSE}
+ml.long2wide <- plot(ml.long2wide.pred)+
+  theme(text=element_text(size=15))+
+  ggtitle(sprintf("Reshaping from long to wide over \nreal numbers, N times", n.folds))+
+  scale_x_log10("N = number of Mean,SD,Length to compute")+
+  scale_y_log10("Computation time (seconds)
+median line, min/max band
+over 10 timings")+
+  facet_null()+
+  scale_fill_manual(values=ml.colors)+
+  scale_color_manual(values=ml.colors)
+
+```
+
+```{r,warning=FALSE,message=FALSE}
+plot(ml.long2wide)
+```
+
+On the other hand, when transforming data from long to wide format, data.table's dcast() function proves to be more efficient.
+
+Although data.table has advantages in terms of writing,reading and Rehaping, pandas in Python still provides a solid framework for data manipulation as seen in the plot.
```

Comment:
writing, reading and reshaping

commas, typo checks, capitalization..check everything when you write please

Reply:
alright